src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -371,146 +371,154 @@ internal unsafe void Start()
-                                // events to skip and notify the user of the rename
-                                ReadOnlySpan<char> newPathRelativeName = events[pairedId].Span.Slice(_fullDirectory.Length);
-                                watcher.NotifyRenameEventArgs(WatcherChangeTypes.Renamed, newPathRelativeName, relativePath);
+                                    // Remove the base directory prefix and add the paired event to the list of
+                                    // events to skip and notify the user of the rename
+                                    ReadOnlySpan<char> newPathRelativeName = ParseEvent(eventPaths[pairedId.GetValueOrDefault()]).Path.Slice(_fullDirectory.Length);
+                                    watcher.NotifyRenameEventArgs(WatcherChangeTypes.Renamed, newPathRelativeName, relativePath);
 
-                                // Create a new list, if necessary, and add the event
-                                if (handledRenameEvents == null)
-                                {
-                                    handledRenameEvents = new List<FSEventStreamEventId>();
+                                    handledRenameEvents = pairedId.GetValueOrDefault();
                                 }
-                                handledRenameEvents.Add(eventIds[pairedId]);
                             }
                         }
                     }
-
-                    ArraySegment<char> underlyingArray;
-                    if (MemoryMarshal.TryGetArray(events[i], out underlyingArray))
-                        ArrayPool<char>.Shared.Return(underlyingArray.Array);
                 }
-
                 this._context = ExecutionContext.Capture();
 
-                void ParseEvents()
+                __ParsedEvent ParseEvent(byte* nativeEventPath)
                 {
-                    for (int i = 0; i < events.Length; i++)
+                    int byteCount = 0;
+                    Debug.Assert(nativeEventPath != null);
+                    byte* temp = nativeEventPath;
+
+                    // Finds the position of null character.
+                    while (*temp != 0)
                     {
-                        int byteCount = 0;
-                        Debug.Assert(eventPaths[i] != null);
-                        byte* temp = eventPaths[i];
+                        temp++;
+                        byteCount++;
+                    }
 
-                        // Finds the position of null character.
-                        while (*temp != 0)
-                        {
-                            temp++;
-                            byteCount++;
-                        }
+                    Debug.Assert(byteCount > 0, "Empty events are not supported");
+                    var tempBuffer = ArrayPool<char>.Shared.Rent(Encoding.UTF8.GetMaxCharCount(byteCount));
+                    Span<char> eventPath = tempBuffer.AsSpan(); ;
```

Comment:
Nit: the `.AsSpan()` isn't necessary... there's an implicit conversion from a `T[]` to a `Span<T>`.